File: tests/cpp/famous_algorithms/test_prims_algorithm.cpp
#include "../../../problems/famous_algorithms/prims_algorithm/cpp/prims_algorithm.h"
#include "../jsontestbase.h"

#include <vector>

using AdjacencyList = std::vector<std::vector<std::vector<int>>>;

class PrimsAlgorithmTest : public JsonTestBase {
public:
  PrimsAlgorithmTest() {
    json_file_path =
        std::string(TEST_CASES_DIR) + "/famous_algorithms/prims_algorithm.json";
  }
};

TEST_F(PrimsAlgorithmTest, TestCases) {
  ASSERT_FALSE(test_cases.empty()) << "Test cases are empty";
  for (const auto &test : test_cases) {
    auto edges = test["edges"].get<AdjacencyList>();

    auto expected = test["expected"].get<AdjacencyList>();
    auto result = primsAlgorithm(edges);

    EXPECT_EQ(result, expected)
        << "Test failed for test cases: " << test.dump(2);
  }
}

File: tests/python/famous_algorithms/prims_algorithm/test_prims_algorithm.py
from pathlib import Path

import matplotlib.pyplot as plt
import networkx as nx

from problems.famous_algorithms.prims_algorithm.python.prims_algorithm import \
    primsAlgorithm
from tests.python.utility import load_test_cases

AdjacencyList = list[list[list[int]]]


def are_permutations(list1: AdjacencyList, list2: AdjacencyList) -> bool:
    if len(list1) != len(list2):
        return False

    sorted_list1 = sorted([sorted(sublist) for sublist in list1])
    sorted_list2 = sorted([sorted(sublist) for sublist in list2])
    return sorted_list1 == sorted_list2


def buildGraph(adj_list: AdjacencyList) -> nx.Graph:
    graph = nx.Graph()
    for i, neighbors in enumerate(adj_list):
        for neighbor, weight in neighbors:
            graph.add_edge(i, neighbor, weight=weight)

    return graph


def drawGraph(graph: nx.Graph, pos: dict, title: str, color: str) -> None:
    nx.draw(graph, pos, with_labels=True, node_color=color, node_size=500, font_size=10)
    labels = nx.get_edge_attributes(graph, "weight")
    nx.draw_networkx_edge_labels(graph, pos, edge_labels=labels)
    plt.title(title)


def visualize_graph(
    edges: AdjacencyList,
    mst: AdjacencyList,
    filename: str,
) -> None:
    root_dir = Path(__file__).resolve().parent
    base_path = root_dir / "prims_algorithm_images"
    base_path.mkdir(parents=True, exist_ok=True)

    original_graph = buildGraph(edges)
    mst_graph = buildGraph(mst)

    pos = nx.spring_layout(original_graph)

    plt.figure(figsize=(12, 6))
    plt.subplot(121)
    drawGraph(original_graph, pos, "Original Graph", "lightblue")
    plt.subplot(122)
    drawGraph(mst_graph, pos, "Minimum Spanning Tree (MST)", "lightgreen")

    output_file = base_path / f"{filename}.png"
    plt.savefig(output_file, format="png", bbox_inches="tight")
    plt.close()
    print(f"Graphs saved to {output_file}")


def test_prims_algorithm(visualize: bool) -> None:

    test_cases = load_test_cases("famous_algorithms/prims_algorithm.json")

    for idx, test in enumerate(test_cases):
        edges: AdjacencyList = test["edges"]
        expected: AdjacencyList = test["expected"]

        result: AdjacencyList = primsAlgorithm(edges)

        assert are_permutations(
            expected, result
        ), f"Test case failed at index {idx}!. Expected: {expected}, got: {result}"

        if visualize:
            visualize_graph(edges, result, f"result_prims_algorithm_{idx}")
